core/platform/src/syscalls_impl.rs

```diff
@@ -1,17 +1,40 @@
 //! Implements `Syscalls` for all types that implement `RawSyscalls`.
 
-use crate::{return_variant, RawSyscalls, Syscalls, YieldType};
+use crate::{RawSyscalls, Syscalls, YieldNoWaitReturn};
+
+mod yield_op {
+    pub const NO_WAIT: u32 = 0;
+    pub const WAIT: u32 = 1;
+}
 
 impl<S: RawSyscalls> Syscalls for S {
     // -------------------------------------------------------------------------
     // Yield
     // -------------------------------------------------------------------------
 
-    fn yield_wait() {
-        Self::raw_yield(YieldType::Wait);
+    fn yield_no_wait() -> YieldNoWaitReturn {
+        unsafe {
+            // flag can be uninitialized because it is not read before the yield
+            // system call, and the kernel promises to only write to it (not
+            // read it).
+            let mut flag = core::mem::MaybeUninit::<YieldNoWaitReturn>::uninit();
+
+            // flag is safe to write a YieldNoWaitReturn to, as guaranteed by
+            // MaybeUninit.
+            Self::yield2(yield_op::NO_WAIT as *mut (), flag.as_mut_ptr() as *mut ());
```

Comment:
> `yield_op::NO_WAIT as *mut ()`

Isn't this... a null pointer?

Reply:
I tried wrapping the `*mut ()` in a struct, which made using it way easier and clearer, but unfortunately you can't use structs in assembly, even if they are `#[repr(transparent)]` :-(